src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala

```diff
@@ -556,7 +556,7 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
         else {
           // match on scrutSym -- converted to an int if necessary -- not on scrut directly (to avoid duplicating scrut)
           Some(BLOCK(
-            ValDef(scrutSym, scrut),
+            ValDef(scrutSym.setInfo(scrutSym.tpe.widen), scrut),
```

Comment:
What happens if the type is not changed here?

Reply:
Then the typechecking that happens in `MatchTransformer#transform` (i.e. when the `translateMatch` there returns) will rewrite `x match {` into `true match {` which isn't an Ident and thus will blows up in CleanUp (look for "these assumptions about the shape of the tree are justified by the codegen in MatchOptimization" in `CleanUp`). 

I don't understand why that type-checking is done there, but if it helps I could unpack the Tree into Block/Match/scrutinee right before it and shift this change to there.

As I see it `setInfo` is what `freshSym` does, so, apart from also dodging some `elimAnonymousClass` and `repeatedToSeq` modifiers, this is pretty much applying that delayed widening now that it matters.

I think it would be nice if singleton types started to work properly in the pattern matching analysis code, but it's tricky business...

Reply:
Maybe it's not too hard to generalize this assumption in CleanUp, let's give that a try.